gnns/rgin.py

```diff
@@ -0,0 +1,135 @@
+    the optimisation of modelling f \circ \phi by a single MLP used in the original
+    GIN formulation.
+
+    We use the following abbreviations in shape descriptions:
+    * V: number of nodes
+    * D: state dimension
+    * L: number of different edge types
+    * E: number of edges of a given edge type
+
+    Arguments:
+        node_embeddings: float32 tensor of shape [V, D], the original representation of
+            each node in the graph.
+        adjacency_lists: List of L adjacency lists, represented as int32 tensors of shape
+            [E, 2]. Concretely, adjacency_lists[l][k,:] == [v, u] means that the k-th edge
+            of type l connects node v to node u.
+        state_dim: Optional size of output dimension of the GNN layer. If not set, defaults
+            to D, the dimensionality of the input. If different from the input dimension,
+            parameter num_timesteps has to be 1.
+        num_timesteps: Number of repeated applications of this message passing layer.
+        activation_function: Type of activation function used.
+        message_aggregation_function: Type of aggregation function used for messages.
+        use_target_state_as_input: Flag indicating if the edge MLP should consume both
+            source and target state (True) or only source state (False).
+        num_MLP_hidden_layers: Number of hidden layers of the MLPs.
+        learn_epsilon: Flag indicating if the value of epsilon should be learned. If
+            False, epsilon defaults to 0.
+
+    Returns:
+        float32 tensor of shape [V, state_dim]
+    """
+    num_nodes = tf.shape(node_embeddings, out_type=tf.int32)[0]
+    if state_dim is None:
+        state_dim = tf.shape(node_embeddings, out_type=tf.int32)[1]
+
+    # === Prepare things we need across all timesteps:
+    activation_fn = get_activation(activation_function)
+    message_aggregation_fn = get_aggregation_function(message_aggregation_function)
+    aggregation_MLP = MLP(out_size=state_dim,
+                          hidden_layers=num_MLP_hidden_layers,
+                          activation_fun=activation_fn)
+    edge_type_to_edge_mlp = []  # MLPs to compute the edge messages
+    edge_type_to_message_targets = []  # List of tensors of message targets
+    for edge_type_idx, adjacency_list_for_edge_type in enumerate(adjacency_lists):
+        with tf.variable_scope("Edge_%i_MLP" % edge_type_idx):
+            edge_type_to_edge_mlp.append(
+                MLP(out_size=state_dim,
+                    hidden_layers=num_MLP_hidden_layers,
+                    activation_fun=activation_fn))
+        edge_type_to_message_targets.append(adjacency_list_for_edge_type[:, 1])
+    # Initialize epsilon: Note that we merge the 1 + \epsilon from the Def. above:
+    if learn_epsilon:
+        epsilon = tf.get_variable("epsilon", shape=(), dtype=tf.float32, initializer=tf.ones_initializer, trainable=True)
+    else:
+        epsilon = 1
```

Comment:
Minor technical thing: maybe allow the user to pass in `epsilon: Optional[float]=None` and when `epsilon is None` imply that it's learnable?

Also, as you note there is a discrepancy here, `epsilon = 1 + ε`  (of Eq 4.1). What about renaming `epsilon` to `epsilon_plus_one` (or something like that)?